jgrapht-io/src/main/java/org/jgrapht/nio/csv/CSVImporter.java

```diff
@@ -0,0 +1,277 @@
+/*
+ * (C) Copyright 2016-2019, by Dimitrios Michail and Contributors.
+ *
+ * JGraphT : a free Java graph-theory library
+ *
+ * See the CONTRIBUTORS.md file distributed with this work for additional
+ * information regarding copyright ownership.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0, or the
+ * GNU Lesser General Public License v2.1 or later
+ * which is available at
+ * http://www.gnu.org/licenses/old-licenses/lgpl-2.1-standalone.html.
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR LGPL-2.1-or-later
+ */
+package org.jgrapht.nio.csv;
+
+import java.io.Reader;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
+import java.util.function.Consumer;
+
+import org.jgrapht.Graph;
+import org.jgrapht.GraphType;
+import org.jgrapht.alg.util.Triple;
+import org.jgrapht.io.CSVFormat;
+import org.jgrapht.io.DefaultAttribute;
+import org.jgrapht.io.GraphImporter;
+import org.jgrapht.io.ImportException;
+import org.jgrapht.nio.BaseConsumerImporter;
+
+/**
+ * Imports a graph from a CSV Format or any other Delimiter-separated value format.
+ * 
+ * <p>
+ * The importer supports various different formats which can be adjusted using the
+ * {@link #setFormat(CSVFormat) setFormat} method. The supported formats are the same CSV formats
+ * used by <a href="https://gephi.org/users/supported-graph-formats/csv-format">Gephi </a>. For some
+ * of the formats, the behavior of the importer can be adjusted using the
+ * {@link #setParameter(org.jgrapht.io.CSVFormat.Parameter, boolean) setParameter} method. See
+ * {@link CSVFormat} for a description of the formats.
+ * </p>
+ * 
+ * <p>
+ * The importer respects <a href="http://www.ietf.org/rfc/rfc4180.txt">rfc4180</a>. The caller can
+ * also adjust the separator to something like semicolon or pipe instead of comma. In such a case,
+ * all fields are unescaped using the new separator. See
+ * <a href="https://en.wikipedia.org/wiki/Delimiter-separated_values">Delimiter- separated
+ * values</a> for more information.
+ * </p>
+ * 
+ * <p>
+ * This importer does not distinguish between {@link CSVFormat#EDGE_LIST} and
+ * {@link CSVFormat#ADJACENCY_LIST}. In both cases it assumes the format is
+ * {@link CSVFormat#ADJACENCY_LIST}.
+ * </p>
+ * 
+ * @see CSVFormat
+ *
+ * @param <V> the graph vertex type
+ * @param <E> the graph edge type
+ * 
+ * @author Dimitrios Michail
+ */
+public class CSVImporter<V, E>
+    extends
+    BaseConsumerImporter<V, E>
```

Comment:
The idea is that all importers behave the same, they notify you about events. 

The `CSVEventDriverImporter` gives you `String` for the vertices and `Triple<String, String, Double>` for the edges and notifies you about additional attributes, etc. 

The `CSVImporter` populates a graph and also notifies you about additional attributes. Even in formats where vertex and/or edge attributes are not present (such as CSV), this is still required since vertices are constructed using vertex suppliers. Thus, the identifier from the input file is not reflected in the actual vertex returned by the vertex supplier. and need to be reported using an event after the vertex is constructed.

Reply:
Got it.  One other question...you introduced the START/END ImportEvents, but currently only the GraphML importers publish them.  Why is that?

Reply:
I added them to the other EventDriven importers as well. I only needed this functionality in the GraphMLImporter, so I forgot about it.